### DISCUSSION TITLE
Extension type missing features

### DISCUSSION THREAD

**Discussion by daniel-white** [dotnet/csharplang#8161] [Language Ideas]
I have a few things I think might be missing from the C# 13's new extension types

1. Control/gating mechanism. It would be nice if there was some method invoked when either implicitly or explicitly converting to an extension type. This would help with a pattern similar to discriminated unions. Operators have a chance to reject a conversion by throwing. Throwing definitely isn't preferred but a `Try...` might be clunky.

```
explicit extension Person for JsonElement
{
     static void Adopt(JsonElement el)
     {
         if (el.GetString("$type") != "person")
         {
             throw new Exception("unable to adopt");
         }
     }
}
```

2. Interfaces. I didn't see it demoed and I doubt extensions have a virtual dispatch. But using interfaces as a constraint to ensure that the extension type has certain members would be beneficial for consistency in a codebase.
```
interface INamed
{
    string Name { get; }
}

implicit extension Person : INamed for JsonElement
{
    // must implement INamed
}

implicit extension Part : INamed for JsonElement
{
    // must implement INamed
}
```

3. Constructors. It would be nice to be able to define a new constructor for a type. IIRC, extensions have access to their own state, so injecting that via a constructor would be beneficial. This could also be another pattern for a gating mechanism.

**Comment by colejohnson66** [comment]
As I understand it, constructors on extensions is something the team is looking at, but it will not be in C# 13. The plan is to release something and improve it later.

It's the same with `readonly` on primary constructors. Release something that works, get feedback, then improve.

**Comment by HaloFour** [comment]
All three are being discussed.  Check out the issue here: https://github.com/dotnet/csharplang/issues/5497

The team is definitely interested in gating/validation of explicit extensions, and there's a whole question/conversation as to how conversions work.

Extension implementation is also a goal as it enables type classes, but it is unlikely to make it for C# 13 given the runtime ramifications.

Extension constructors have been mentioned recently too, although I've not heard anything concrete as to how they would behave.

**Comment by TahirAhmadov** [comment]
I think gating/validation for explicit extensions is very important for that feature to be useful. Absent such a mechanism, it looks a lot like a pit of failure to me.

**Reply by HaloFour** [reply]
We were chatting about that on the [Dotnet Evolution Discord](https://aka.ms/dotnet-discord) yesterday and it does seem like something that the team is _very_ interested in.  It's still under discussion but it sounds like:

1. Explicit extensions would get an explicit conversion operator by default, but not an implicit conversion operator.
2. That explicit conversion operator can be instead provided by the developer, which would allow for validation that can throw if the conversion were to fail.
3. An implicit conversion operator can also be provided if the developer chooses.
4. There is interest in a custom `is` operator for pattern matching that can inspect the underlying type, so given (syntax is 100% spaghetti I invented):
    ```cs
    public explicit extension PersonNode for JsonObject {
        public static bool operator is(JsonObject x) {
            return x.GetString("$type") == "person";
        }
    }

    JsonObject json = ...;
    switch (json) {
        case PersonNode node:
            ...
    }
    ```

**Reply by Mrxx99** [reply]
The "is operator" is what would make explicit extensions useful for me.

**Comment by KennethHoff** [comment]
With explicit extensions being kind of ( but not really) functional/TypeScript-like", it would be nice if something akin to the following TypeScript code was possible.


```ts
type JsonString = string & { _type: "json }; // branded type - this would be the explicit extension.


function isJson(value: string): value is JsonString {
    // Logic to determine whether it's a Json string, returning true if it is, and false if it isn't.
}


const str = ....; // type: string

if(isJson(str)) {
   // str is now type: JsonString
}


```

**Reply by colejohnson66** [reply]
The idea of `is` operators would allow this:

```csharp
string json = GetJsonString();
if (json is JsonString jsonString)
    UseJsonString(jsonString);

public implicit extension JsonString for string
{
    public static bool operator is(string str) => true;
}
```

**Reply by HaloFour** [reply]
Again, syntax is 100% spaghetti, but it sounds like the team is interested in something akin to the following:

```cs
public explicit extension JsonString for string {
    public static bool operator is(string val) {
        // Logic to determine whether it's a JsonString, returning true if it is, and false if it isn't.
    }

    public static operator explicit JsonString(string val) {
        // Logic to determine whether it's a JsonString, throwing if it isn't.
    }
}

var str = "...";
if (str is JsonString jstr) {
    // use jstr here
}

// optionally
JsonString jstr = (JsonString) str; // calls explicit conversion operator

```

**Reply by TahirAhmadov** [reply]
Do we really need the conversion operator to be user definable? By definition, extensions cannot have any state, so there is nothing else to do when "converting" to an extension. Therefore, just the `is` operator should be sufficient.

```cs
public explicit extension JsonString for string {
    public static bool operator is(string val) {
        // Logic to determine whether it's a JsonString, returning true, and false if it isn't.
    }
}

var str = "...";
if (str is JsonString jstr) { // call "is" operator, if it returns true, proceed to "then" block
    // use jstr here
}

// optionally
JsonString jstr = (JsonString) str; // calls "is" operator, if it returns false, throw, otherwise, success
```

For implicit extensions, I think we don't even need - in fact, we can't really have - an "is" operator (or any validation mechanism).

**Reply by HaloFour** [reply]
> Therefore, just the `is` operator should be sufficient.

I don't, for the same reason `int.TryParse` and `int.Parse` are separate.  The former can bail quickly and return `false` as soon as it detects that the type does not match, whereas the latter can throw an exception with details as to why the conversion is illegal.

**Reply by TahirAhmadov** [reply]
There is a huge difference - parsing a string to int involves doing work. There is by definition _no work_ to wrap a value into an explicit extension - or am I missing anything?

**Reply by HaloFour** [reply]
> There is a huge difference - parsing a string to int involves doing work. There is by definition no work to wrap a value into an explicit extension - or am I missing anything?

So does this:

```cs
public explicit extension NumericString for string { ... }
```

So does parsing into a bit of JSON or XML to determine if the document adheres to a certain schema or contains certain attributes/elements.  Pretty much any scenario where an extension might conditionally apply I would want to have a mechanism to understand why.

**Reply by TahirAhmadov** [reply]
OK I think I need to demonstrate exactly where I'm coming from. 
```cs
public explicit extension PersonNode for JsonObject {
    public static bool operator is(JsonObject x) {
        return x.GetString("$type") == "person";
    }
    public static operator explicit PersonNode (JsonObject val) {
        // basically, the logic here is one-for-one copy of the "is" operator
        // just with a throw instead of return false - why do it?
        if( x.GetString("$type") != "person") throw new InvalidCastException(); // or whatever

        return ....; // what do we do here?
    }
}
```
PS. Another problem is that if we let people define their own explicit operators, they can throw any exception they want, which means it becomes impossible to standardize the ecosystem on one exception type, say, a hypothetical `InvalidExtensionException`. It's better to stick to `return false` and let the language do the throwing, for consistency.

**Reply by TahirAhmadov** [reply]
> So does parsing into a bit of JSON or XML to determine if the document adheres to a certain schema or contains certain attributes/elements. Pretty much any scenario where an extension might conditionally apply I would want to have a mechanism to understand why.

Yes but that parsing is only for validation purposes in both cases. You cannot take the results of said parsing and store it anywhere **because extensions do not carry any state of their own**. This is the difference from `int.Parse`; its result is an actual `int` value which is distinct and different from the `string` input. With extensions, there is no new value we're producing - period.

**Reply by TahirAhmadov** [reply]
PPS. And I didn't even mention the most obvious pitfall here - what's stopping the developer from having one logic in `is operator` and a different logic in the `explicit operator`? Absolutely nothing. So the `is operator` can check for condition X, and the `explicit operator` can check for condition Y and throw exception P, and also check for condition Z and throw exception Q, and the consumers of said extension are left wondering why things work completely differently when they should work identically.

**Reply by TahirAhmadov** [reply]
PPPS. If anything, an argument can be made that for flexibility, we want to allow the extension authors to throw different exception types depending on which incompatibility exists between a given instance and the attempted extension. In that case, the `is` check should call said operator and catch all exceptions and return false in the catch.
```cs
public explicit extension PersonNode for JsonObject {
    public static operator is (JsonObject val) { // void return, not bool
        if( x.GetString("$type") != "person") throw new InvalidTypeException(); 
        if( x.GetString("$first") == "") throw new FirstNameMissingException(); 
    }
}
JsonObject json = ...;
switch (json) {
    case PersonNode person: 
       // becomes: try { PersonNode.is(json); } catch { goto companyCase; } person = new WhateverStructWrapperThingie(json);
        ...
       break;
    case CompanyNode company:
}

if(json is PersonNode) { ... }
// becomes: try { PersonNode.is(json); } catch { goto elseCase; } ... elseCase:

var person = (PersonNode)json; // becomes: PersonNode.is(json); person = new WhateverStructWrapperThingie(json);
```

**Reply by HaloFour** [reply]
>  In that case, the `is` check should call said operator and catch all exceptions and return false in the catch.

Which would make it ridiculously expensive.

If you don't want to do it both ways that's your choice.  You can have one operator call the other and call it a day.  Heck, I'd even be fine if that's what the compiler would emit by default, as long as I could override that myself where I felt appropriate.  We don't even know exactly what shape these things will definitely take.  However, if was explicitly mentioned that wanting to be able to throw different exceptions with specific details as to why an explicit conversion didn't work was a desirable feature, and that also makes it completely unsuitable for simple `Try` conditions.

**Reply by TahirAhmadov** [reply]
That's fine; basically, the default is a) the obviously faster `bool is`, and the language emits `throw new InvalidExtensionException()` basing on result of that `bool is`; b) if you want to provide more info, you can define the throwing `void is` operator in addition; or c) if you want to do the `void is` only, then the language emits the `bool is` which has the `try catch` in it - perhaps with a warning/info that this code path always incurs the cost of try-catch.
PS. I'm still concerned that if we allow both operators, the potential for inconsistency is a problem, but I suppose we can't really stop people from shooting themselves in the foot.

**Reply by HaloFour** [reply]
IMO, custom pattern operators should never throw.

**Reply by TahirAhmadov** [reply]
```cs
public explicit extension PersonNode for JsonObject {
    public static Exception? operator is (JsonObject val) {
        if( x.GetString("$type") != "person") return InvalidTypeException.Default; // allocations are for losers
        if( x.GetString("$first") == "") return new FirstNameMissingException(); 
        return null;
    }
}
```
Bingo!!

PS. The singleton exception idea is silly, because there is no stack trace, but maybe we can do magic and return singletons when calling in an `is` context and `throw new` in a cast context, at runtime, or with a transpiler? Really thinking out of the box here, don't quote me on any of this.

**Reply by HaloFour** [reply]
I dislike that approach.  It's still conflating two concerns in the name of simplicity, and I think the result is more complex for it.

This also makes the assumption that custom patterns are something specific to extensions, and I don't think they should be.  The fact that extensions can't (currently) have state shouldn't dictate how an orthogonal feature is designed.

**Reply by TahirAhmadov** [reply]
These are not two concerns though. Fundamentally the `is` and the "cast" **should work identically.** My approach allows that to happen without copying and pasting code, etc. It's DRY and performant.

Regarding "custom patterns" - we'll cross that bridge when we get there. We're talking about extensions here, which is a specific feature.

If you need state, then you don't need an extension - you need a custom struct which is composed of the underlying value and then you can have any state and extra functionality you want on it. An SG (like my composition SG) can "bring" the underlying API area to your custom struct.

**Reply by TahirAhmadov** [reply]
```cs
public explicit extension PersonNode for JsonObject {
    public static operator is (JsonObject val) {
        if( x.GetString("$type") != "person") throw new InvalidTypeException();
        if( x.GetString("$first") == "") throw new FirstNameMissingException(); 
    }
    // compiler emits automatically and it cannot be overridden:
    public static bool operator is (JsonObject val) {
        if( x.GetString("$type") != "person") return false;
        if( x.GetString("$first") == "") return false;
        return true;
    }
}
```
Done.
PS. This won't work probably if another method is called as part of the validation logic - the compiler cannot go inside of it to transpile anything. :(

**Reply by HaloFour** [reply]
> These are not two concerns though. Fundamentally the is and the "cast" should work identically. My approach allows that to happen without copying and pasting code, etc. It's DRY and performant.

It conflates a conditional test with an unconditional conversation for the sake of jamming both concerns into a single method, and as a result ends up with a single signature and behavior that makes it feel alien to both use cases.  It also makes `is` unnecessarily more expensive and larger as it needs to do bookkeeping it otherwise wouldn't need to do.

> Regarding "custom patterns" - we'll cross that bridge when we get there. We're talking about extensions here, which is a specific feature.

We're talking about custom patterns as members of extensions.  They're not going to be designed twice for the extensions and non-extensions.

> If you need state, then you don't need an extension

Extensions can't currently carry state because there is no cheap and easy way to accomplish that in the runtime _today_.  Don't confuse that with a design decision that states that extensions should never be capable of carrying state, nor should the design of these features preclude that as a possibility in the future.

**Reply by TahirAhmadov** [reply]
> It conflates a conditional test with an unconditional conversation for the sake of jamming both concerns into a single method, 

If you have:
```cs
object o = Foo();
// o is not null

if(o is string s) { Bar(s); }

string s = (string)o; Bar(s);
```
Would you want there to be any chance that the `o is string` and `(string)o` use different criteria for success?
Obviously not.
The "sake" is not jamming it into a single method - it's to have an approach which reflects what conceptually makes sense. The reason here is that conceptually, the cast and the `is` **have** to work identically. Having one method for both is just the DRY method of achieving it and reducing number of potential bugs.

> and as a result ends up with a single signature and behavior that makes it feel alien to both use cases. 

"Feel alien" is a subjective opinion, I can't answer that in any way other than it "feels perfectly human" to me 🤣 

> It also makes `is` unnecessarily more expensive and larger as it needs to do bookkeeping it otherwise wouldn't need to do.

Not true, with my transpilation approach at https://github.com/dotnet/csharplang/discussions/8161#discussioncomment-9586800, both approaches have optimal performance.

> We're talking about custom patterns as members of extensions. They're not going to be designed twice for the extensions and non-extensions.

No we're not. We're taking about validation for explicit extensions. Custom patterns is a completely different feature.

> Extensions can't currently carry state because there is no cheap and easy way to accomplish that in the runtime _today_. Don't confuse that with a design decision that states that extensions should never be capable of carrying state, nor should the design of these features preclude that as a possibility in the future.

I disagree. Extensions with state are not extensions.

**Reply by HaloFour** [reply]
> Would you want there to be any chance that the o is string and (string)o use different criteria for success?

This is already the case with custom conversion operators.

> Having one method for both is just the DRY method of achieving it and reducing number of potential bugs.

Having symmetric members is common throughout C# and the runtime.  That's why `operator ==` doesn't delegate to `Equals` and why `operator !=` doesn't delegate to negating the result of `operator ==`.  The individual concerns should be addressed individually.  If they can shame some implementation, or you choose to have one call the other, that is your choice.

> No we're not. We're taking about validation for explicit extensions. Custom patterns is a completely different feature.

Through custom patterns, which are a completely different feature and should be designed as such.

> I disagree. Extensions with state are not extensions.

Based on what?  That sounds quite arbitrary.  Scala extension types (implicit classes) can have state just fine.

**Reply by HaloFour** [reply]
At this point we're talking past each other, and it doesn't matter, as neither of our opinions is going to drive the shape of these features.  The LDM has expressed their interest in this space, I suggest we see what they bring to the table.

**Reply by TahirAhmadov** [reply]
> This is already the case with custom conversion operators.

Can you demonstrate a code example of what you mean here?

> That's why operator == doesn't delegate to Equals

While I think the entire equality situation is messed up and needs a clean up, I wouldn't expect `==` and `Equals` to be the same thing "just because". On the other hand ...

> operator != doesn't delegate to negating the result of operator ==.

To me, that's a WTF in the language. Even if there is a need to do something differently for `!=` (say, for performance reasons), it should be optional and an edge case, the current approach of it being mandatory is dumbfounding to me.

> Through custom patterns, which are a completely different feature and should be designed as such.

Talking about "conflating", custom patterns by definition do not align well with the extension casting. I don't see how these 2 things can be addressed at the same time. Extension validation is different.

>  Scala extension types (implicit classes) can have state just fine.

I'm not a Scala expert but their docs say that "implicit classes" are obsoleted and only supported for backwards compatibility and the approach going forward is extension methods, which don't have state. This leads me to believe they made some mistakes with them - perhaps adding state was one such mistake? 😉

**Reply by HaloFour** [reply]
> Can you demonstrate a code example of what you mean here?

```cs
var b = new B();
if (b is C) { } // nope
var c = (C) b; // yep (fixed)

class B { }

class C {
    public static explicit operator C(B b) => new();
}

```

> Talking about "conflating", custom patterns by definition do not align well with the extension casting. I don't see how these 2 things can be addressed at the same time. Extension validation is different.

If it uses `is`, it's custom pattern.  There's no reason to design a completely different flavor specific to extensions, or specific to extension validation.  This is the opposite of conflating, it's designing two orthogonal features that can be combined and composed together.  That's how most of C# is designed.

**Reply by TahirAhmadov** [reply]
I think you meant: `var c = (C)b; // yep`
There is a difference here - this is a conversion operator, whereas for explicit extensions, we're answering a completely different question; it's "can this value be extended in a certain way", as opposed to conversion operators which answer "how do we take X and produce a completely different Y" out of it. 
In terms of looking for things in the language that extension validation is consistent with, it would be casting (not converting) and `as`/`is` operators for types. The question there is, "can X be cast to Y". And there, the underlying behavior of all 3 - cast, `as` and `is` - is exactly the same, and it should be the same with extension validation.

> If it uses is, it's custom pattern. 

Except it also uses the cast syntax. We can't proclaim that because it has `is`, it's a custom pattern, full stop, while completely ignoring the need to make the "cast" syntax work, which raises the question of how both of these approaches work, and we answer that int the only obviously correct way - that the "cast" and the `is` syntaxes should work the same way.
In other words, it's much more important to address the question of how to make the casting and the `is` work congruently for extension validation, than to look for non-existent similarities with custom patterns (for things like case insensitive string comparisons, etc.)

**Reply by HaloFour** [reply]
> I think you meant: `var c = (C)b; // yep`

oops!  😅 I fixed it in the original post.

> this is a conversion operator, whereas for explicit extensions, we're answering a completely different question

I disagree with the idea of designing something completely bespoke to extensions.

>  Except it also uses the cast syntax. We can't proclaim that because it has `is`, it's a custom pattern, full stop, while completely ignoring the need to make the "cast" syntax work, which raises the question of how both of these approaches work, and we answer that int the only obviously correct way - that the "cast" and the is syntaxes should work the same way.

Patterns aren't casts, and none of the patterns for custom/active patterns that are already on this repo treats them as such.  Active patterns in most languages are just an identifier.

https://github.com/dotnet/csharplang/issues/4131
https://github.com/dotnet/csharplang/issues/1047

Again, we're going in circles here and have a fundamental disagreement.  I don't think there's much value in continuing this here.  If you want to bring it to the Discord channel, that's fine.

**Reply by TahirAhmadov** [reply]
> I disagree with the idea of designing something completely bespoke to extensions.

Why, though? It is a question which is specific to extensions. What's the problem with answering an extension-specific question with an extension-specific answer? In fact, I think it's exactly the right way to go.

> Patterns aren't casts, and none of the patterns for custom/active patterns that are already on this repo treats them as such.

Exactly. With extension validation, we want both the `is` and the cast. Which is exactly why custom/active patterns are irrelevant here.

Yes I agree, let's leave it at this. I'm very interested in hearing others' opinions on this, too

**Reply by HaloFour** [reply]
> Why, though? It is a question which is specific to extensions

None of these questions are specific to extensions.

**Reply by Mrxx99** [reply]
What about having these 3 options as a user:
a) Don't overwrite any operator -> default behavior
b) only overwrite is operator -> compiler generated explicit operator will use the is operator and throw a standardized exception on false
c) Overwrite both is and explicit operator -> you are on your own, but have more flexibility

Not allowed:
Overwrite only explicit operator

a) and b) should already be enough for > 90% of use cases and c) could be used if more flexibility is required, here the user should (but doesn't have to) make sure himself that the behavior matches between the two operators

**Reply by CyrusNajmabadi** [reply]
@Mrxx99 That's my internal mental model currently for approaching this.

**Reply by HaloFour** [reply]
I think that model makes sense.  My only concern is that it would introduce an immediate inconsistency with bringing custom patterns to non-extensions.  IMO, an explicit conversion without a custom pattern would be fine, but you don't get the ability to use that extension with pattern matching.

**Reply by TahirAhmadov** [reply]
Regarding the `explicit operator`, I still can't understand what do we do in the "success" scenario - do we need a special syntax to do the "conversion"?
```cs
public explicit extension PersonNode for JsonObject {
    public static bool operator is(JsonObject x) {
        return x.GetString("$type") == "person";
    }
    public static operator explicit PersonNode (JsonObject val) {
        // basically, the logic here is one-for-one copy of the "is" operator
        // just with a throw instead of return false - why do it?
        if( x.GetString("$type") != "person") throw new InvalidCastException(); // or whatever

        return ....; // what do we do here? <<<<============
    }
}
```
`return ....; // what do we do here?` - I'm not being facetious, what is being returned?
`return val;` - special rule inside the `explicit operator` to "commit"
`return new PersonNode(val);` - some "shadow" ctor - is it only valid inside `explicit operator`, what happens if you try outside of it? Can't create an extension ctor with the same signature anymore.
PS. What if this `explicit operator` calls another method, how do we allow to return the `extension` in a random static method, for example?

**Reply by HaloFour** [reply]
That could be a concern.  Random spaghetti, but I would suggest that perhaps the language allow for implicit conversion to the extension type from anywhere within that extension type.  This may pose a problem if the extension also declares an implicit conversion operator, and perhaps within said implicit conversion operator the language allows assignment to the extension type directly.

**Reply by HaloFour** [reply]
Also, if custom pattern operators are to be a general purpose feature and work for non-extensions, they would need to support the target type as an output from the operator in some manner.  Perhaps it could follow the `Try`/`out` pattern:

```cs
public static bool operator is(JsonObject json, out PersonNode node) {
    if (json.GetString("$type") == "person") {
        node = json;
        return true;
    }
    return false;
}
```

**Reply by TahirAhmadov** [reply]
Sorry but the word "shoehorning" comes to mind.

**Reply by HaloFour** [reply]
> Sorry but the word "shoehorning" comes to mind.

No more than `Try` methods are today.  This is quite literally the same design as suggested by language design team members for such a feature in the past.  It's how destructuring works, and a `bool`-returning `Deconstruct` method was also one of the possible designs for custom patterns.

**Reply by daniel-white** [reply]
Please let’s set details aside. My post is more about the spirit of having something and not get stuck in the minutiae hence I called my example method Adopt.

**Reply by HaloFour** [reply]
> Please let’s set details aside

Details aside, the team has expressed being very interested in this space.  So we'll have to wait and see what kind of design they bring to the table in order to enable this.  :)

**Reply by TahirAhmadov** [reply]
> No more than `Try` methods are today. This is quite literally the same design 

https://github.com/dotnet/csharplang/discussions/8161#discussioncomment-9586303
The `Try` methods address a completely different use case. Again, "shoehorning".

> It's how destructuring works, and a `bool`-returning `Deconstruct` method was also one of the possible designs for custom patterns.

Again, the `Deconstruct` takes something and transforms it into something else. The extension validation is **different.**
Literally the only thing in common here is the word "is".

![6m8ivk](https://github.com/dotnet/csharplang/assets/16783987/e5159e69-69bb-4e5d-a3ef-715b30f0f99f)

Sorry couldn't help myself :)

**Reply by TahirAhmadov** [reply]
If we want to allow syntax like `string s = "123"; if(s is int i) { ... }` in the future, that's fine; however, other than `is`, the actual operation is different, and therefore has to use a different mechanism.

In fact, if we were to imagine a hypothetical extension custom pattern system in the future:
```cs
// leave aside the question of something common like this might be part of BCL:
implicit extension StringEx for string
{
  public static bool operator is(out int res) { ... }
}
```
Then it means we have to steer very clear of using `is` for extension validation, because it creates some (not necessarily a deal breaker, but still) ambiguity. 
I'm actually thinking it would be best to do:
```cs
public explicit extension PersonNode for JsonObject 
{
  // Validate is a recognized method name, similar to Deconstruct (exact name doesn't matter as much)
  public static bool Validate(JsonObject x) 
  { 
    return x.GetString("$type") == "person";
  }
  public static bool operator is (out Customer res) 
  {
    if(this.GetString("$relationship") == "customer")
    {
      res = new Customer(...) { ... };
      return true;
    }
    return false;
  }
}
class Customer(...) { ... }

JsonObject json = ...;
if(json is PersonNode person) // relies on the "Validate" method
{ 
  // ...
  if(person is Customer customer) // uses the "is" operator
  { ... }
}
```

**Reply by CyrusNajmabadi** [reply]
> Then it means we have to steer very clear of using is for extension validation, because it creates some (not necessarily a deal breaker, but still) ambiguity.

What's the ambiguity?

**Reply by HaloFour** [reply]
> the actual operation is different, and therefore has to use a different mechanism.

I don't see any difference between those two.  In fact, I find their similarity to be a virtue.  Both involve validation and converting the type.  The fact that one type is a runtime type and the other type is an extension type that is elided by the compiler is an implementation detail.

**Reply by TahirAhmadov** [reply]
> What's the ambiguity?

```cs
public explicit extension PersonNode for JsonObject 
{
  public static bool operator is(JsonObject x) 
  {
        return x.GetString("$type") == "person";
  }
  public static bool operator is (out Customer res) 
  {
    if(this.GetString("$relationship") == "customer")
    {
      res = new Customer(...) { ... };
      return true;
    }
    return false;
  }
}
```
Above, we have to `is`es, and one does something completely different than the other one. Technically, it's probably possible to disambiguate between them because of the different signatures, etc., but visually they look too similarly for being aimed at addressing 2 different scenarios.
Having said that, if it's deemed not to be a big issue WRT the visual confusion, it wouldn't be the end of the world to have both like in the code snippet above. Less than ideal, but not terrible.

**Reply by TahirAhmadov** [reply]
> Both involve validation and converting

There is no "converting", though. That's the part I don't seem to be able to get across to you. The only thing we have with extensions is validation. 
Validation and converting != validation only.

**Reply by HaloFour** [reply]
> There is no "converting", though.

I don't see why there needs to be a distinction, certainly not one that necessitates designing two features that do the same thing.  I see performing an `is` pattern and coming out the other end with another type, whether that be an extension or a different runtime type.  The fact that it is actually the same type and the extension is elided is, again, an implementation detail.

**Reply by CyrusNajmabadi** [reply]
I don't understand the ambiguity.  What's ambiguous in the above?

**Reply by TahirAhmadov** [reply]
> I don't see why there needs to be a distinction,

There doesn't need to be a distinction - there is a distinction. I'm not the one who doesn't want there to be logic to do conversion - by design, there can be no conversion logic. 

> I see performing an is pattern and coming out the other end with another type

At the call site, sure - I'm not against using the `is` operator for both concepts. It's on the implementation side where there clearly is a difference, and said difference has to be recognized, not massaged into an incompatible signature.

**Reply by HaloFour** [reply]
> There doesn't need to be a distinction - there is a distinction.

I feel like this is something that you want to attribute here that is not necessary.  The fact that there isn't an actual transformation of the data doesn't seem necessary, anymore so that such an operator would apply to something that behaves as a wrapper struct.
